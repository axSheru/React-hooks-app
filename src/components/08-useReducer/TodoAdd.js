import React from 'react';
import { useForm } from '../../hooks/useForm';
import './styles.css';

export const TodoAdd = ({ handleAddTodo,  }) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    });

    const handleSubmit = ( e ) => {
        e.preventDefault();
    
        if ( description.trim().length <= 1 ) return;
    
        const newTodo = {
            id: new Date().getTime(),
            desc: description.trim(),
            done: false,
        };
    
        handleAddTodo( newTodo );
    
        reset();
    };

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={ handleSubmit } >
                <input
                    type="text"
                    name="description"
                    className='form-control'
                    placeholder='Debo de....'
                    autoComplete='off'
                    value={ description }
                    onChange={ handleInputChange }
                />
                <div className="d-grid gap-2">
                    <button
                        className='btn btn-outline-primary mt-1'
                        type='submit'
                    >
                        Agregar
                    </button>
                </div>
            </form>
        </>
    )
}
