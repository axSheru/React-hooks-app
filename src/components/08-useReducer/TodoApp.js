import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';
import './styles.css';

const init = () => {
    /* return [{
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false,
    }]; */

    return JSON.parse(localStorage.getItem( 'todos' )) || [];
};

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer( todoReducer, [], init);

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    });

    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) );
    }, [todos]);

    const handleDelete = ( todoId ) => {

        const action = {
            type: 'delete',
            payload: todoId,
        };

        dispatch( action );

    };
    

    const handleSubmit = ( e ) => {
        e.preventDefault();

        if ( description.trim().length <= 1 ) return;

        console.log('Nueva tarea.');

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };

        const action = {
            type: 'add',
            payload: newTodo,
        };

        dispatch( action );

        reset();
    };

    return (
        <div>
            <h1>TodoApp ({ todos.length })</h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <ul className='list-group list-group-flush'>
                        {
                            todos.map( ( todo, i ) => (
                                <li
                                    key={ todo.id }
                                    className='list-group-item'
                                >
                                    <p className='text-center'>{ i + 1 }. { todo.desc }</p>
                                    <button
                                        className='btn btn-outline-danger'
                                        onClick={ () => handleDelete( todo.id ) }
                                    >
                                        Borrar
                                    </button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar TODO</h4>
                    <hr />

                    <form onSubmit={ handleSubmit } >
                        <input
                            type="text"
                            name="description"
                            className='form-control'
                            placeholder='Debo de....'
                            autoComplete='off'
                            value={ description }
                            onChange={ handleInputChange }
                        />
                        <div className="d-grid gap-2">
                            <button
                                className='btn btn-outline-primary mt-1'
                                type='submit'
                            >
                                Agregar
                            </button>
                        </div>
                    </form>
                </div>
            </div>

        </div>
    )
};
