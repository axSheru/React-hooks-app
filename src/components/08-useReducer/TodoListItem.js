import React from 'react';
import './styles.css';

export const TodoListItem = ({ todo, i, handleToggle, handleDelete }) => {
    return (
        <li
            key={ todo.id }
            className='list-group-item'
        >
            <p
                className={ `text-center ${ todo.done && 'complete' }` }
                onClick={ () => handleToggle( todo.id ) }
            >
                { i + 1 }. { todo.desc }
            </p>
            <button
                className='btn btn-outline-danger'
                onClick={ () => handleDelete( todo.id ) }
            >
                Borrar
            </button>
        </li>
    )
}
