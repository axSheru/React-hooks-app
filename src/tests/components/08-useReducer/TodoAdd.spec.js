import { shallow } from "enzyme";
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";

describe('Pruebas en el componente TodoAdd.', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow( <TodoAdd
        handleAddTodo={ handleAddTodo }
    /> );

    test('debe de hacer match con el snapshot.', () => {

        expect( wrapper ).toMatchSnapshot();
        
    });

});