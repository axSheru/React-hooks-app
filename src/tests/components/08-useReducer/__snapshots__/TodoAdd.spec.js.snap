// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente TodoAdd. debe de hacer match con el snapshot. 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Debo de...."
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-outline-primary mt-1"
        type="submit"
      >
        Agregar
      </button>
    </div>
  </form>
</Fragment>
`;
